File: #0139.word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {

        int n = s.size();
        vector<bool> v(n + 1, false);
        v[n] = true;

        for (int i = n - 1; i >= 0; --i) {
            for (int j = i + 1; j <= n; ++j) {
                if (v[j] && find(wordDict.begin(), wordDict.end(), s.substr(i, j - i)) != wordDict.end()) {
                    v[i] = true;
                    break;
                }
            }
        }

        return v[0];
    }
};

File: #0139.word-break.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
    const dp = new Array(s.length + 1);
    dp[0] = true;

    for (let i = 0; i < s.length; i++) {
        if (!dp[i]) {
            continue;
        }

        wordDict.forEach(w => {
            if (s.slice(i, i + w.length) === w) {
                dp[i + w.length] = true;
            }
        });
    }
    return !!dp[s.length];
};
